### DISCUSSION TITLE
UseEffect Problem

### DISCUSSION THREAD

**Discussion by Lolly1150** [vercel/next.js#40907] [Help]
I have socket connection in useEffect i but it connect more time
![immagine](https://user-images.githubusercontent.com/86166828/192314763-7cfabb18-9fd8-4874-a637-dfc282f399aa.png)

**Comment by icyJoseph** [comment]
Got a repo, or reproduction demo to look at?

Either, `StrictMode` is making it connect several times, or you have some dependency that's changing more than you want to in `useEffect`, or perhaps you don't have dependencies at all?

Could you at least share some pseudo-code of the Component using the `useEffect`?

**Reply by Lolly1150** [reply]
```ts
  useEffect(() => {
    socket.on("connect", async () => {
      console.log("[WS] Connected");
    });
    socket.on("error", (err: any) => {
      console.log("[WS] Error: ", err);
    });
    socket.on("rate-limit", (data) => {
      console.log(data);
      socket.disconnect();
    });
    socket.on("disconnect", (reason) => {
      console.log(`[WS] Disconnected: ${reason}`);
      socket.disconnect();
    });
    socket.emit("user");
    socket.on("user", async (data: any) => {
      return setUser(data);
    });
  }, [socket]);
```

**Reply by icyJoseph** [reply]
One thing already is that this effect doesn't clean-up the socket.

What library are you using?

And, now the follow up question is, how do you instantiate the socket? It seem that the way you have it creates a new instance on every render pass.

**Reply by Lolly1150** [reply]
socket-io client

**Reply by Lolly1150** [reply]
I have found fix removing socket on array dep

**Reply by icyJoseph** [reply]
Yes, well that's not necessarily a fix. You need to do as they show in their docs:

```tsx
import React, { useState, useEffect } from 'react';
import io from 'socket.io-client';

const socket = io(); // socket in this case is global to the module

function App() {
  const [isConnected, setIsConnected] = useState(socket.connected);
  const [lastPong, setLastPong] = useState(null);

  useEffect(() => {
    socket.on('connect', () => {
      setIsConnected(true);
    });

    socket.on('disconnect', () => {
      setIsConnected(false);
    });

    socket.on('pong', () => {
      setLastPong(new Date().toISOString());
    });

    return () => {
      // this part is specially important
      socket.off('connect');
      socket.off('disconnect');
      socket.off('pong');
    };
  }, []); // they don't include the dependency here because it is defined for the module

  const sendPing = () => {
    socket.emit('ping');
  }

  return (
    <div>
      <p>Connected: { '' + isConnected }</p>
      <p>Last pong: { lastPong || '-' }</p>
      <button onClick={ sendPing }>Send ping</button>
    </div>
  );
}

export default App;
```

I guess you were invoking `const socket = io()` as part of the React Component body, and not outside as in their example?

**Reply by Lolly1150** [reply]
for automatization of send ping and i must be add in return socket.off('my event name emmited')

**Reply by Lolly1150** [reply]
What is socket.off?

**Reply by icyJoseph** [reply]
`socket.off` is to disconnect and stop listening, for example if your component un-mounts, but you don't clean-up, you'll still try to update state for an unmounted component.

In dev mode React prints a warning when that happens, telling you that you have a memory leak.

**Reply by Lolly1150** [reply]
if i emit user i must be add all events is socket.off
